mat/triangular_test.go

```diff
@@ -513,23 +513,109 @@ func TestCopySymIntoTriangle(t *testing.T) {
 	}
 }
 
-func BenchmarkTriSum1000(b *testing.B) { triSumBench(b, 1000) }
-
 var triSumForBench float64
 
-func triSumBench(b *testing.B, size int) {
-	a := randTriDense(size)
-	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		triSumForBench = Sum(a)
+func BenchmarkTriSum(b *testing.B) {
+	rnd := rand.New(rand.NewSource(1))
+	benchmarks := []struct {
+		name string
+		size int
+	}{
+		{name: "BenchmarkTriSum100", size: 100},
+		{name: "BenchmarkTriSum200", size: 200},
+		{name: "BenchmarkTriSum400", size: 400},
+		{name: "BenchmarkTriSum800", size: 800},
+		{name: "BenchmarkTriSum1600", size: 1600},
+	}
+	for _, bm := range benchmarks {
+		a := randTriDense(bm.size, rnd)
+		b.Run(bm.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				triSumForBench = Sum(a)
+			}
+		})
+	}
+}
+
+var triProductForBench *TriDense
+
+func BenchmarkTriMul(b *testing.B) {
+	rnd := rand.New(rand.NewSource(1))
+	benchmarks := []struct {
+		name string
+		size int
+	}{
+		{name: "BenchmarkMulTri100", size: 100},
+		{name: "BenchmarkMulTri200", size: 200},
+		{name: "BenchmarkMulTri400", size: 400},
+		{name: "BenchmarkMulTri800", size: 800},
+		{name: "BenchmarkMulTri1600", size: 1600},
+	}
+	for _, bm := range benchmarks {
+		triProductForBench = NewTriDense(bm.size, Upper, nil)
+		a := randTriDense(bm.size, rnd)
+		c := randTriDense(bm.size, rnd)
+		b.Run(bm.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				triProductForBench.MulTri(a, c)
+			}
+		})
+	}
+}
+
+func BenchmarkTriMulDiag(b *testing.B) {
+	rnd := rand.New(rand.NewSource(1))
+	benchmarks := []struct {
+		name string
+		size int
+	}{
+		{name: "BenchmarkMulTri100", size: 100},
+		{name: "BenchmarkMulTri200", size: 200},
+		{name: "BenchmarkMulTri400", size: 400},
+		{name: "BenchmarkMulTri800", size: 800},
+		{name: "BenchmarkMulTri1600", size: 1600},
+	}
+	for _, bm := range benchmarks {
+		triProductForBench = NewTriDense(bm.size, Upper, nil)
+		a := randTriDense(bm.size, rnd)
+		c := randDiagDense(bm.size, rnd)
+		b.Run(bm.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				triProductForBench.MulTri(a, c)
+			}
+		})
+	}
+}
+
+func BenchmarkTriMul2Diag(b *testing.B) {
+	rnd := rand.New(rand.NewSource(1))
+	benchmarks := []struct {
+		name string
+		size int
+	}{
+		{name: "BenchmarkMulTri100", size: 100},
+		{name: "BenchmarkMulTri200", size: 200},
+		{name: "BenchmarkMulTri400", size: 400},
+		{name: "BenchmarkMulTri800", size: 800},
+		{name: "BenchmarkMulTri1600", size: 1600},
+	}
+	for _, bm := range benchmarks {
+		triProductForBench = NewTriDense(bm.size, Upper, nil)
+		a := randDiagDense(bm.size, rnd)
+		c := randDiagDense(bm.size, rnd)
+		b.Run(bm.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				triProductForBench.MulTri(a, c)
+			}
+		})
 	}
 }
 
-func randTriDense(size int) *TriDense {
+func randTriDense(size int, source *rand.Rand) *TriDense {
```

Comment:
Use `rnd` rather than `source`. We'd normally call a source `src`, but it is a different thing to a `rnd` anyway; one being the source of randomness and the other being a value that can give specifically defined random distribution randomness.

Reply:
ok now, hopefully...